### PR TITLE
✨ (api) Allow multiple authentication methods (oidc & basic)

### PR DESCRIPTION
**Pull request by Leobouloc**
## Purpose

Currently, an instance of Ralph LRS has a fixed auth method (basic auth and OIDC). The suggestion is to allow multiple auth methods. An example use case could be: "an LMS writes to Ralph LRS via basic Auth; while data analysts fetch from Ralph using OIDC".

## Proposal

It is proposed to change `RALPH_RUNSERVER_AUTH_BACKEND` to `RALPH_RUNSERVER_AUTH_BACKENDS` and to make it accept comma separated lists as values `RALPH_RUNSERVER_AUTH_BACKENDS=basic,oidc`.

The auth method for incoming requests will be chosen based on header information (`basic`, `oidc`)

- [x] implement suggestion
- [x] test that auth methods work individually and simultaneously

### CODE REVIEW

File: src/ralph/conf.py [hunk 1]
```diff
@@ -133,6 +133,36 @@ class Config:  # pylint: disable=missing-class-docstring # noqa: D106
...
+    BASIC = "Basic"
+    OIDC = "OIDC"
+
+
+class AuthBackends(str):
+    """Model representing a list of authentication backends."""
+
+    @classmethod
+    def __get_validators__(cls):  # noqa: D105
+        """Checks whether the value is a comma separated string or a tuple representing
```
Comment (wilbrdt) [inline src/ralph/conf.py:148]:
```suggestion
        """Check whether the value is a comma separated string or a tuple representing
```

File: tests/test_conf.py [hunk 2]
```diff
@@ -9,6 +9,11 @@
 from ralph.conf import CommaSeparatedTuple, Settings, settings
 from ralph.exceptions import ConfigurationException
 
+# import os
+# def test_env_dist(fs, monkeypatch):
+#     fs.create_file(".env", contents=os.read("../.env.dist"))
+#     Settings()
+
```
Comment (wilbrdt) [inline tests/test_conf.py:16]: Don't forget about this too!

**Review by wilbrdt — COMMENTED** [review]
Alt text

File: tests/test_conf.py (hunk 2 above)
Comment (Leobouloc) [reply tests/test_conf.py:16]: oups. yes this is for a separate PR

File: CHANGELOG.md [hunk 3]
```diff
@@ -51,6 +51,8 @@ have an authority field matching that of the user
 - Helm: improve volumes and ingress configurations
 - API: Add `RALPH_LRS_RESTRICT_BY_SCOPE` option enabling endpoint access
   control by user scopes
+- API: Variable `RUNSERVER_AUTH_BACKEND` becomes `RUNSERVER_AUTH_BACKENDS`, and
```
[the remark this answers is not in the document]
Comment (Leobouloc) [reply CHANGELOG.md:54]: changed !

File: src/ralph/api/auth/__init__.py [hunk 4]
```diff
@@ -1,11 +1,48 @@
...
+    basic_auth_user=Depends(get_basic_auth_user),
+    oidc_auth_user=Depends(get_oidc_user),
+):
+    """Authenticate user with any allowed method, using credentials in the header."""
+    if AuthBackend.BASIC not in settings.RUNSERVER_AUTH_BACKENDS:
+        basic_auth_user = None
+    if AuthBackend.OIDC not in settings.RUNSERVER_AUTH_BACKENDS:
+        oidc_auth_user = None
+
+    if basic_auth_user is not None:
```
[the remark this answers is not in the document]
Comment (Leobouloc) [reply src/ralph/api/auth/__init__.py:22]: changed

File: src/ralph/api/auth/__init__.py [hunk 5]
```diff
@@ -1,11 +1,48 @@
...
 from ralph.api.auth.oidc import get_oidc_user
-from ralph.conf import settings
+from ralph.conf import AuthBackend, settings
+
+
+def get_authenticated_user(
+    security_scopes: SecurityScopes = SecurityScopes([]),
+    basic_auth_user=Depends(get_basic_auth_user),
+    oidc_auth_user=Depends(get_oidc_user),
+):
```
[the remark this answers is not in the document]
Comment (Leobouloc) [reply src/ralph/api/auth/__init__.py:15]: changed

File: src/ralph/api/auth/__init__.py [hunk 6]
```diff
@@ -1,11 +1,48 @@
...
 from ralph.api.auth.basic import get_basic_auth_user
 from ralph.api.auth.oidc import get_oidc_user
-from ralph.conf import settings
+from ralph.conf import AuthBackend, settings
+
+
+def get_authenticated_user(
+    security_scopes: SecurityScopes = SecurityScopes([]),
+    basic_auth_user=Depends(get_basic_auth_user),
+    oidc_auth_user=Depends(get_oidc_user),
```
[the remark this answers is not in the document]
Comment (Leobouloc) [reply src/ralph/api/auth/__init__.py:14]: changed

File: src/ralph/api/auth/oidc.py [hunk 7]
```diff
@@ -66,7 +66,7 @@ def discover_provider(base_url: AnyUrl) -> Dict:
         )
         raise HTTPException(
             status_code=status.HTTP_401_UNAUTHORIZED,
-            detail="Could not validate credentials",
+            detail="Could not validate credentials ABU",
```
[the remark this answers is not in the document]
Comment (Leobouloc) [reply src/ralph/api/auth/oidc.py:69]: changed for TODO to test this

File: src/ralph/api/auth/oidc.py [hunk 8]
```diff
@@ -88,14 +88,13 @@ def get_public_keys(jwks_uri: AnyUrl) -> Dict:
         )
         raise HTTPException(
             status_code=status.HTTP_401_UNAUTHORIZED,
-            detail="Could not validate credentials",
+            detail="Could not validate credentials ABA",
```
[the remark this answers is not in the document]
Comment (Leobouloc) [reply src/ralph/api/auth/oidc.py:91]: changed for TODO to test this

File: tests/test_conf.py (hunk 2 above)
Comment (Leobouloc) [reply tests/test_conf.py:16]: see: https://github.com/openfun/ralph/issues/462

File: src/ralph/api/auth/__init__.py [hunk 9]
```diff
@@ -1,11 +1,48 @@
...
+    """Authenticate user with any allowed method, using credentials in the header."""
+    if AuthBackend.BASIC not in settings.RUNSERVER_AUTH_BACKENDS:
+        basic_auth_user = None
+    if AuthBackend.OIDC not in settings.RUNSERVER_AUTH_BACKENDS:
+        oidc_auth_user = None
+
+    if basic_auth_user is not None:
+        user = basic_auth_user
+        auth_method = "Basic"
+    elif oidc_auth_user is not None:
```
[the remark this answers is not in the document]
Comment (Leobouloc) [reply src/ralph/api/auth/__init__.py:25]: changed

>

File: src/ralph/api/auth/__init__.py [hunk 10]
```diff
@@ -1,11 +1,48 @@
...
+        user = oidc_auth_user
+        auth_method = "Bearer"
+    else:
+        raise HTTPException(
+            status_code=status.HTTP_401_UNAUTHORIZED,
+            detail="Invalid authentication credentials",
+            headers={
+                "WWW-Authenticate": ",".join(
+                    [val.value for val in settings.RUNSERVER_AUTH_BACKENDS]
```
[the remark this answers is not in the document]
Comment (Leobouloc) [reply src/ralph/api/auth/__init__.py:37]: I don't have a strong opinion. I leave this open, if someone else wants to comment.

File: .env.dist [hunk 11]
```diff
@@ -116,9 +116,9 @@ RALPH_BACKENDS__HTTP__LRS__STATEMENTS_ENDPOINT=/xAPI/statements
 
 # RALPH_CONVERTER_EDX_XAPI_UUID_NAMESPACE=
 
-# LRS API
+# LRS API 
 
-RALPH_RUNSERVER_AUTH_BACKEND=basic
+RALPH_RUNSERVER_AUTH_BACKENDS=Basic
```
Comment (jmaupetit) [inline .env.dist:121]: If we choose enum values to be lower case:

```suggestion
RALPH_RUNSERVER_AUTH_BACKENDS=basic
```

File: src/ralph/conf.py [hunk 12]
```diff
@@ -133,6 +133,36 @@ class Config:  # pylint: disable=missing-class-docstring # noqa: D106
     timeout: float
 
 
+class AuthBackend(Enum):
+    """Model for valid authentication methods."""
+
+    BASIC = "Basic"
+    OIDC = "OIDC"
```
Comment (jmaupetit) [inline src/ralph/conf.py:140]: Shouldn't all values be upper or lower case but not a mix?

File: src/ralph/conf.py [hunk 13]
```diff
@@ -133,6 +133,36 @@ class Config:  # pylint: disable=missing-class-docstring # noqa: D106
     timeout: float
 
 
+class AuthBackend(Enum):
```
Comment (jmaupetit) [inline src/ralph/conf.py:136]:
```suggestion
class AuthBackend(StrEnum):
```

File: src/ralph/api/auth/basic.py [hunk 14]
```diff
@@ -145,15 +138,14 @@ def get_basic_auth_user(
     except StopIteration:
         # next() gets the first item in the enumerable; if there is none, it
         # raises a StopIteration error as it is out of bounds.
-        logger.warning(
+        logger.info(
```
Comment (jmaupetit) [inline src/ralph/api/auth/basic.py:141]: IMO the WARNING level was fine

File: src/ralph/api/auth/basic.py [hunk 15]
```diff
@@ -121,18 +119,13 @@ def get_basic_auth_user(
...
         HTTPException
     """
     if not credentials:
-        logger.error("The basic authentication mode requires a Basic Auth header")
-        raise HTTPException(
-            status_code=status.HTTP_401_UNAUTHORIZED,
-            detail="Could not validate credentials",
-            headers={"WWW-Authenticate": "Basic"},
-        )
+        logger.info("No credentials were found for Basic auth")
```
Comment (jmaupetit) [inline src/ralph/api/auth/basic.py:127]: We will need to check if this is not too noisy and eventually switch to the DEBUG level for this.

File: src/ralph/api/auth/basic.py [hunk 16]
```diff
@@ -145,15 +138,14 @@ def get_basic_auth_user(
...
             "User %s tried to authenticate but this account does not exists",
             credentials.username,
         )
         hashed_password = None
-    except AuthenticationError as exc:
-        raise HTTPException(
-            status_code=status.HTTP_403_FORBIDDEN, detail=str(exc)
-        ) from exc
+    except AuthenticationError:
+        logger.info("Error while authenticating using Basic auth")
```
Comment (jmaupetit) [inline src/ralph/api/auth/basic.py:147]: If we want this log to be informative, we need more context to identify which user performed the action. WDYT?

File: src/ralph/api/auth/oidc.py [hunk 17]
```diff
@@ -109,17 +108,25 @@ def get_oidc_user(
...
+
     if auth_header is None or "Bearer" not in auth_header:
-        logger.error("The OpenID Connect authentication mode requires a Bearer token")
-        raise HTTPException(
-            status_code=status.HTTP_401_UNAUTHORIZED,
-            detail="Could not validate credentials",
-            headers={"WWW-Authenticate": "Bearer"},
+        logger.info(
+            "Not using OIDC auth. The OpenID Connect authentication mode requires a "
+            "Bearer token"
```
Comment (jmaupetit) [inline src/ralph/api/auth/oidc.py:115]: IMO this is for debug purpose. This would be too noisy and uninformative in production.

**Review by SergioSim — COMMENTED** [review]
Looks great :+1:  Have only minor questions/suggestions)

image

File: src/ralph/api/__init__.py [hunk 18]
```diff
@@ -43,6 +43,7 @@ def filter_transactions(
     )
 
 app = FastAPI()
+
```
[the remark this answers is not in the document]
Comment (Leobouloc) [reply src/ralph/api/__init__.py:44]: yes

File: src/ralph/conf.py [hunk 19]
```diff
@@ -133,6 +133,36 @@ class Config:  # pylint: disable=missing-class-docstring # noqa: D106
...
+    OIDC = "OIDC"
+
+
+class AuthBackends(str):
+    """Model representing a list of authentication backends."""
+
+    @classmethod
+    def __get_validators__(cls):  # noqa: D105
+        """Check whether the value is a comma separated string or a tuple representing
+        an AuthBackend."""
```
[the remark this answers is not in the document]
Comment (Leobouloc) [reply src/ralph/conf.py:149]: changed for all instances in code

File: src/ralph/api/auth/basic.py (hunk 14 above)
Comment (Leobouloc) [reply src/ralph/api/auth/basic.py:141]: reverted to warning

File: src/ralph/api/auth/basic.py (hunk 15 above)
Comment (Leobouloc) [reply src/ralph/api/auth/basic.py:127]: set it to DEBUG

File: src/ralph/api/auth/basic.py (hunk 16 above)
Comment (Leobouloc) [reply src/ralph/api/auth/basic.py:147]: Reverted to the original exception, following @SergioSim 's comment

File: src/ralph/api/auth/basic.py [hunk 20]
```diff
@@ -145,15 +138,14 @@ def get_basic_auth_user(
...
+        return None
```
[the remark this answers is not in the document]
Comment (Leobouloc) [reply src/ralph/api/auth/basic.py:148]: This makes sense. Reverted to the original exception.

File: src/ralph/api/auth/oidc.py [hunk 21]
```diff
@@ -109,17 +108,25 @@ def get_oidc_user(
     Raises:
         HTTPException
     """
+
     if auth_header is None or "Bearer" not in auth_header:
```
[the remark this answers is not in the document]
Comment (Leobouloc) [reply src/ralph/api/auth/oidc.py:113]: Yes ! `According to HTTP specification header names are case-insensitive`

Changed.

File: src/ralph/api/auth/oidc.py (hunk 17 above)
Comment (Leobouloc) [reply src/ralph/api/auth/oidc.py:116]: agreed. changed.

File: src/ralph/conf.py (hunk 13 above)
Comment (Leobouloc) [reply src/ralph/conf.py:136]: made the change with `class AuthBackend(str, Enum):`

File: src/ralph/conf.py [hunk 22]
```diff
@@ -133,6 +133,36 @@ class Config:  # pylint: disable=missing-class-docstring # noqa: D106
...
 
+class AuthBackend(Enum):
+    """Model for valid authentication methods."""
+
+    BASIC = "Basic"
+    OIDC = "OIDC"
+
+
+class AuthBackends(str):
+    """Model representing a list of authentication backends."""
```
[the remark this answers is not in the document]
Comment (Leobouloc) [reply src/ralph/conf.py:144]: oups

File: .env.dist (hunk 11 above)
Comment (Leobouloc) [reply .env.dist:156]: changed this

File: src/ralph/conf.py (hunk 12 above)
Comment (Leobouloc) [reply src/ralph/conf.py:140]: all lower then ?

File: src/ralph/conf.py [hunk 23]
```diff
@@ -133,6 +133,36 @@ class Config:  # pylint: disable=missing-class-docstring # noqa: D106
...
+    def __get_validators__(cls):  # noqa: D105
+        """Check whether the value is a comma separated string or a tuple representing
+        an AuthBackend."""
+
+        def validate(
+            value: Union[AuthBackend, Tuple[AuthBackend], List[AuthBackend]]
+        ) -> Tuple[AuthBackend]:
+            """Check whether the value is a comma separated string or a list/tuple."""
+            if isinstance(value, (tuple, list)):
+                return tuple(AuthBackend(value))
```
[the remark this answers is not in the document]
Comment (Leobouloc) [reply src/ralph/conf.py:157]: added

File: src/ralph/api/auth/__init__.py [hunk 24]
```diff
@@ -1,11 +1,51 @@
...
+    security_scopes: SecurityScopes = SecurityScopes([]),
+    basic_auth_user: Optional[AuthenticatedUser] = Depends(get_basic_auth_user),
+    oidc_auth_user: Optional[AuthenticatedUser] = Depends(get_oidc_user),
+) -> AuthenticatedUser:
+    """Authenticate user with any allowed method, using credentials in the header."""
+    if AuthBackend.BASIC not in settings.RUNSERVER_AUTH_BACKENDS:
+        basic_auth_user = None
+    if AuthBackend.OIDC not in settings.RUNSERVER_AUTH_BACKENDS:
+        oidc_auth_user = None
+
```
[the remark this answers is not in the document]
Comment (Leobouloc) [reply src/ralph/api/auth/__init__.py:22]: Done

File: src/ralph/conf.py [hunk 25]
```diff
@@ -133,6 +133,44 @@ class Config:  # noqa: D106
...
+                    AuthBackend(value.lower()) for value in auth_backends.split(",")
+                )
+
+            if isinstance(auth_backends, AuthBackend):
+                return (auth_backends,)
+
+            if isinstance(auth_backends, (tuple, list)):
+                return tuple(auth_backends)
+
+            raise TypeError("Invalid comma-separated list")
```
Comment (Leobouloc) [inline src/ralph/conf.py:169]: modified since last review

File: tests/test_conf.py [hunk 26]
```diff
@@ -54,10 +60,33 @@ def test_conf_comma_separated_list_with_valid_values(value, expected, monkeypatc
...
+def test_conf_auth_backend(value, is_valid, expected, monkeypatch):
+    """Test the AuthBackends data type with valid and invalid values."""
+    if is_valid:
+        assert next(AuthBackends.__get_validators__())(value) == expected
+        monkeypatch.setenv("RALPH_RUNSERVER_AUTH_BACKENDS", "".join(value))
+        reload(conf)
+        assert conf.settings.RUNSERVER_AUTH_BACKENDS == expected
+    else:
+        with pytest.raises(ValueError, match="'notvalid' is not a valid AuthBackend"):
+            next(AuthBackends.__get_validators__())(value)
```
Comment (Leobouloc) [inline tests/test_conf.py:87]: modified since last review

File: src/ralph/api/auth/oidc.py [hunk 27]
```diff
@@ -110,13 +109,12 @@ def get_oidc_user(
     Raises:
         HTTPException
     """
-    if auth_header is None or "Bearer" not in auth_header:
-        logger.error("The OpenID Connect authentication mode requires a Bearer token")
-        raise HTTPException(
-            status_code=status.HTTP_401_UNAUTHORIZED,
-            detail="Could not validate credentials",
-            headers={"WWW-Authenticate": "Bearer"},
+    if auth_header is None or "bearer" not in auth_header.lower():
```
[the remark this answers is not in the document]
Comment (Leobouloc) [reply src/ralph/api/auth/oidc.py:112]: Not sure. 

`auth_header`, is an instance of `HTTPBearer`. The `HTTPBearer` class checks for `if scheme.lower() != "bearer":` right before returning `bearer` which is then shipped as is.

**Review by wilbrdt — APPROVED** [review]
LGTM! :tada:

**Pull request merged**